### PR TITLE
[WIP] Streams 2.4.18 update

### PR DESCRIPTION
**Pull request by Arkatufus**
Scala port of https://github.com/akka/akka/pull/22033, https://github.com/akka/akka/pull/22657, and https://github.com/akka/akka/pull/22647

### CODE REVIEW

**Comment by Silv3rcircl3** [comment]
part of #2601

File: src/core/Akka.Streams.Tests/IO/FileSinkSpec.cs [hunk 1]
```diff
@@ -209,6 +210,23 @@ public void SynchronousFileSink_should_allow_overriding_the_dispatcher_using_Att
...
+        public void SynchronousFileSink_should_write_single_line_to_a_file_from_lazy_sink()
+        {
+            this.AssertAllStagesStopped(() => {
+                TargetFile(f => {
+                    var completion = Source.From(_testByteStrings);
+                    completion.RunWith(Sink.LazySink<ByteString, Task<IOResult>>(
+                            b => Task.FromResult(FileIO.ToFile(f)),
+                            () => Task.FromResult(new IOResult(0, new Result<NotUsed>(NotUsed.Instance)))
+                        ), _materializer);
+                    completion.MapMaterializedValue();
```
Comment (Silv3rcircl3) [inline src/core/Akka.Streams.Tests/IO/FileSinkSpec.cs]: this needs to be `completion.MapMaterializedValue(list => list.SelectMany(x => x));` not sure about the actual type of `list` but `flatMap` means `SelectMany`

File: src/core/Akka.Streams.Tests/IO/FileSinkSpec.cs [hunk 2]
```diff
@@ -209,6 +210,23 @@ public void SynchronousFileSink_should_allow_overriding_the_dispatcher_using_Att
...
+                    
+                });
```
Comment (Silv3rcircl3) [inline src/core/Akka.Streams.Tests/IO/FileSinkSpec.cs]: task needs to be awaited + `CheckFileContent`

File: src/core/Akka.Streams/Implementation/Sinks.cs [hunk 3]
```diff
@@ -938,6 +939,15 @@ public override void OnPush()
                     _stage._sinkFactory(element)
                         .ContinueWith(t => callback(Result.FromTask(t)),
                             TaskContinuationOptions.ExecuteSynchronously);
+                    SetHandler(_stage.In, new LambdaInHandler(
+                            onPush: () => { },
+                            onUpstreamFinish: () =>
+                            {
+                                SetKeepGoing(true);
```
Comment (Silv3rcircl3) [inline src/core/Akka.Streams/Implementation/Sinks.cs]: Move this into `GotCompletionEvent`

File: src/core/Akka.Streams/Implementation/Sinks.cs [hunk 4]
```diff
@@ -984,7 +994,7 @@ private sealed class SubSource : SubSourceOutlet<TIn>
             {
                 private readonly Logic _logic;
                 private readonly LazySink<TIn, TMat> _stage;
-                private bool _completed;
+                internal bool Completed;
```
Comment (Silv3rcircl3) [inline src/core/Akka.Streams/Implementation/Sinks.cs]: move `Completed` into the `Logic` class

File: src/core/Akka.Streams/Implementation/Sinks.cs [hunk 5]
```diff
@@ -984,7 +994,7 @@ private sealed class SubSource : SubSourceOutlet<TIn>
...
 
                 public SubSource(Logic logic, TIn firstElement) : base(logic, "LazySink")
```
Comment (Silv3rcircl3) [inline src/core/Akka.Streams/Implementation/Sinks.cs]: you get the `Logic` as parameter and can use `_logic.Completed` in the two places below.

File: src/core/Akka.Streams/Implementation/Sinks.cs [hunk 6]
```diff
@@ -1005,7 +1015,7 @@ public SubSource(Logic logic, TIn firstElement) : base(logic, "LazySink")
                         onUpstreamFinish: () =>
```
Comment (Silv3rcircl3) [inline src/core/Akka.Streams/Implementation/Sinks.cs]: replace with `_logic.GotCompletionEvent()`

**Comment by Arkatufus** [comment]
Partially resolved, moving to a new branch.

**Pull request closed without merging**